crates/uv-interpreter/src/discovery.rs

```diff
@@ -0,0 +1,1134 @@
+                        version.is_none() || version.is_some_and(|version|
+                            version.matches_version(toolchain.python_version())
+                        )
+                    )
+                    .map(|toolchain| (InterpreterSource::ManagedToolchain, toolchain.executable()))
+                )
+                .map_err(Error::from)
+            ).flatten_ok()
+        ).into_iter().flatten()
+    )
+    // (4) The search path
+    .chain(
+        sources.contains(InterpreterSource::SearchPath).then(move ||
+            python_executables_from_search_path(version)
+            .map(|path| Ok((InterpreterSource::SearchPath, path))),
+        ).into_iter().flatten()
+    )
+    // (5) The `py` launcher (windows only)
+    // TODO(konstin): Implement <https://peps.python.org/pep-0514/> to read python installations from the registry instead.
+    .chain(
+        (sources.contains(InterpreterSource::PyLauncher) && cfg!(windows)).then(||
+            std::iter::once(
+                py_list_paths()
+                .map(|entries|
+                    // We can avoid querying the interpreter using versions from the py launcher output unless a patch is requested
+                    entries.into_iter().filter(move |entry|
+                        version.is_none() || version.is_some_and(|version|
+                            version.has_patch() || version.matches_major_minor(entry.major, entry.minor)
+                        )
+                    )
+                    .map(|entry| (InterpreterSource::PyLauncher, entry.executable_path))
+                )
+                .map_err(Error::from)
+            ).flatten_ok()
+        ).into_iter().flatten()
+    )
+}
+
+/// Lazily iterate over Python executables in the `PATH`.
+///
+/// The [`VersionRequest`] is used to determine the possible Python interpreter names, e.g.
+/// if looking for Python 3.9 we will look for `python3.9` in addition to the default names.
+///
+/// Executables are returned in the search path order, then by specificity of the name, e.g.
+/// `python3.9` is preferred over `python3`.
+///
+/// If a `version` is not provided, we will only look for default executable names e.g.
+/// `python3` and `python` — `python3.9` and similar will not be included.
+fn python_executables_from_search_path(
+    version: Option<&VersionRequest>,
+) -> impl Iterator<Item = PathBuf> + '_ {
+    // `UV_TEST_PYTHON_PATH` can be used to override `PATH` to limit Python executable availability in the test suite
+    let search_path =
+        env::var_os("UV_TEST_PYTHON_PATH").unwrap_or(env::var_os("PATH").unwrap_or_default());
+    let possible_names = version.unwrap_or(&VersionRequest::Default).possible_names();
+
+    // Split and iterate over the paths instead of using `which_all` so we can
+    // check multiple names per directory while respecting the search path order
+    let search_dirs: Vec<_> = env::split_paths(&search_path).collect();
```

Comment:
Nit: Do you need to collect here? You immediately `into_iter`, can you just continue the chain?

Reply:
```
error[E0515]: cannot return value referencing local variable `search_path`
   --> crates/uv-interpreter/src/discovery.rs:268:5
    |
268 |       env::split_paths(&search_path)
    |       ^                ------------ `search_path` is borrowed here
    |  _____|
    | |
269 | |         .into_iter()
270 | |         .filter(|dir| dir.is_dir())
271 | |         .flat_map(move |dir| {
...   |
302 | |                 )
303 | |         })
    | |__________^ returns a value referencing data owned by the current function
    |
    = help: use `.collect()` to allocate the iterator
```